File: MoonEngine-Root/ECS/src/main/java/com/greentree/engine/moon/ecs/ClassSet.java
package com.greentree.engine.moon.ecs;

import java.io.Externalizable;
import java.io.IOException;
import java.io.ObjectInput;
import java.io.ObjectOutput;
import java.util.Collection;
import java.util.HashSet;
import java.util.Iterator;
import java.util.Objects;

import com.greentree.commons.action.observable.TypedObjectObservable;
import com.greentree.commons.action.observer.type.TypedObjectAction;
import com.greentree.commons.util.collection.OneClassSet;
import com.greentree.commons.util.iterator.IteratorUtil;

@SuppressWarnings({"unchecked", "deprecation"})
public class ClassSet<E> implements Iterable<E>, Externalizable {

	private transient final TypedObjectAction<Class<? extends E>, E> addAction = new TypedObjectAction<>();
	private final OneClassSet<E> components = new OneClassSet<>();
	private transient final TypedObjectAction<Class<? extends E>, E> removeAction = new TypedObjectAction<>();

	public void clear() {
		try(var lock = lock()) {
			for(E c : components)
				lock.remove((Class<? extends E>) c.getClass());
		}
	}


	public boolean contains(Class<? extends E> componentClass) {
		return components.containsClass(componentClass);
	}

	public boolean contains(E component) {
		return components.contains(component);
	}


	@Override
	public String toString() {
		return "ClassSet " + components;
	}


	public <T extends E> T get(Class<T> componentClass) {
		Objects.requireNonNull(componentClass);
		if(!contains(componentClass))
			throw new IllegalArgumentException("componenet "+componentClass.getName()+" not contains in " + this);
		return components.get(componentClass);
	}

	public TypedObjectObservable<? extends Class<? extends E>, ? extends E> getAddAction() {
		return addAction;
	}

	public TypedObjectObservable<? extends Class<? extends E>, ? extends E> getRemoveAction() {
		return removeAction;
	}
	public boolean isEmpty() {
		return components.isEmpty();
	}

	@Override
	public Iterator<E> iterator() {
		return components.iterator();
	}

	public LockClassSet<E> lock() {
		return new LockClassSet<>(this);
	}

	@Override
	public void readExternal(ObjectInput in) throws IOException, ClassNotFoundException {
		var size = in.readInt();
		try(final var lock = lock()) {
			while(size-- > 0) {
				final var c = (E) in.readObject();
				lock.add(c);
			}
		}
	}
	public int size() {
		return components.size();
	}

	@Override
	public void writeExternal(ObjectOutput out) throws IOException {
		out.writeInt(components.size());
		for(var c : this)
			out.writeObject(c);
	}

	protected Iterable<? extends Class<? extends E>> getClassRequired(Class<? extends E> cls) {
		return IteratorUtil.empty();
	}

	public static final class LockClassSet<E> implements AutoCloseable {

		private final Collection<E> addedComponent = new HashSet<>(), removedComponent = new HashSet<>();

		private final ClassSet<E> _this;
		private LockClassSet(ClassSet<E> _this) {
			this._this = _this;
		}

		private static boolean containsClass(Iterable<?> iter, Class<?> cls) {
			for(var i : iter)
				if(cls.isInstance(i))
					return true;
			return false;
		}

		public void add(E component) {
			Objects.requireNonNull(component);
			if(removedComponent.contains(component))
				removedComponent.remove(component);
			else {
				final var cls = component.getClass();
				if(addedComponent.contains(component))
					throw new IllegalArgumentException(cls + " aready added");
				addedComponent.add(component);
			}
		}

		@Override
		public void close() {
			for(E c : addedComponent) {
				final var cls = (Class<? extends E>) c.getClass();
				if(_this.contains(cls))
					throw new IllegalArgumentException("componenet "+cls.getName()+" already added to " + _this);
				final var requiredComponents = _this.getClassRequired(cls);
				for(var rc : requiredComponents)
					if(!_this.contains(rc) && !containsClass(addedComponent, rc))
						throw new IllegalArgumentException("add componet " + c + " required " + rc.getName() + " to "
								+ _this + ", also added " + addedComponent);
			}
			_this.components.addAll(addedComponent);
			for(E c : addedComponent) {
				final var cls = (Class<? extends E>) c.getClass();
				_this.addAction.event(cls, c);
			}
			for(E c : removedComponent) {
				final var cls = (Class<? extends E>) c.getClass();
				_this.removeAction.event(cls, c);
			}
			_this.components.removeAll(removedComponent);
		}

		public <T extends E> T get(Class<T> componentClass) {
			return _this.get(componentClass);
		}

		public <T extends E> T remove(Class<T> componentClass) {
			final var component = _this.get(componentClass);
			if(addedComponent.contains(component))
				addedComponent.remove(component);
			else {
				if(removedComponent.contains(component))
					throw new IllegalArgumentException(componentClass + " aready remove");
				removedComponent.add(component);
			}
			return component;
		}

	}

}

File: MoonEngine-Root/ECS/src/main/java/com/greentree/engine/moon/ecs/ClassSet.kt
package com.greentree.engine.moon.ecs

import com.greentree.commons.action.observable.TypedObjectObservable
import com.greentree.commons.action.observer.type.TypedObjectAction
import com.greentree.commons.util.collection.OneClassSet
import com.greentree.commons.util.iterator.IteratorUtil
import java.io.Externalizable
import java.io.IOException
import java.io.ObjectInput
import java.io.ObjectOutput
import java.util.*

@Suppress("deprecation")
open class ClassSet<E : Any> : Iterable<E>, Externalizable {

	@Transient
	private val addAction = TypedObjectAction<Class<out E>, E>()

	@Transient
	private val removeAction = TypedObjectAction<Class<out E>, E>()
	private val components = OneClassSet<E>()

	fun clear() {
		lockWith {
			for (c in components)
				remove(c::class.java)
		}
	}

	val isEmpty
		get() = components.isEmpty()

	operator fun contains(componentClass: Class<out E>): Boolean {
		return components.containsClass(componentClass)
	}

	operator fun contains(component: E): Boolean {
		return components.contains(component)
	}

	override fun toString(): String {
		return "ClassSet $components"
	}

	operator fun <T : E> get(componentClass: Class<T>): T {
		require(contains(componentClass)) { "componenet " + componentClass.name + " not contains in " + this }
		return components.get(componentClass)
	}

	fun getAddAction(): TypedObjectObservable<out Class<out E>, out E> {
		return addAction
	}

	fun getRemoveAction(): TypedObjectObservable<out Class<out E>, out E> {
		return removeAction
	}

	override fun iterator(): MutableIterator<E> {
		return components.iterator()
	}

	private fun <R> lockWith(function: LockClassSet<E>.() -> R): R {
		return lock().use(function)
	}

	private fun <R> lockUse(function: (LockClassSet<E>) -> R): R {
		return lock().use(function)
	}

	fun lock(): LockClassSet<E> {
		return LockClassSet(this)
	}

	@Throws(IOException::class, ClassNotFoundException::class)
	override fun readExternal(objectInput: ObjectInput) {
		var size = objectInput.readInt()
		lockUse { lock ->
			while (size-- > 0) {
				val c = objectInput.readObject() as E
				lock.add(c)
			}
		}
	}

	fun size(): Int {
		return components.size
	}

	@Throws(IOException::class)
	override fun writeExternal(out: ObjectOutput) {
		out.writeInt(components.size)
		for (c in this) out.writeObject(c)
	}

	class LockClassSet<E : Any>(private val _this: ClassSet<E>) : AutoCloseable {

		private val addedComponent: MutableCollection<E> = HashSet()
		private val removedComponent: MutableCollection<E> = HashSet()

		fun add(component: E) {
			if (removedComponent.contains(component))
				removedComponent.remove(component)
			else {
				val cls = component::class
				require(!addedComponent.contains(component)) { "$cls aready added" }
				addedComponent.add(component)
			}
		}

		override fun close() {
			for (c in addedComponent) {
				val cls = c.javaClass as Class<out E>
				require(!_this.contains(cls)) { "component " + cls.name + " already added to " + _this }
				val requiredComponents = _this.getClassRequired(cls)
				for (rc in requiredComponents) require(!(!_this.contains(rc) && !containsClass(addedComponent, rc))) {
					("add component " + c + " required " + rc.name + " to "
							+ _this + ", also added " + addedComponent)
				}
			}
			_this.components.addAll(addedComponent)
			for (c in addedComponent) {
				val cls = c.javaClass as Class<out E>
				_this.addAction.event(cls, c)
			}
			for (c in removedComponent) {
				val cls = c.javaClass as Class<out E>
				_this.removeAction.event(cls, c)
			}
			_this.components.removeAll(removedComponent)
		}

		fun <T : E> remove(componentClass: Class<T>): T {
			val component = _this.get(componentClass)
			if (addedComponent.contains(component)) addedComponent.remove(component) else {
				require(!removedComponent.contains(component)) { "$componentClass already remove" }
				removedComponent.add(component)
			}
			return component
		}
	}

	protected open fun getClassRequired(cls: Class<out E>): Iterable<Class<out E>> {
		return IteratorUtil.empty()
	}
}

private inline fun containsClass(iter: Iterable<*>, cls: Class<*>): Boolean {
	return iter.any { cls.isInstance(it) }
}